numpy/core/src/multiarray/iterators.c

```diff
@@ -1242,232 +1242,183 @@ PyArray_Broadcast(PyArrayMultiIterObject *mit)
+static NPY_INLINE PyObject*
+multiiter_wrong_number_of_args(void)
+{
+    return PyErr_Format(PyExc_ValueError,
+                        "Need at least 1 and at most %d "
+                        "array objects.", NPY_MAXARGS);
+}
+
+/*
+ * Common implementation for all PyArrayMultiIterObject constructors.
+ * This function takes a pointer to an array of at most NPY_MAXARGS
+ * PyObject pointers, NULL terminated unless completely full, which
+ * must either hold array_like objects or multi iterators.
  */
-NPY_NO_EXPORT PyObject *
-PyArray_MultiIterFromObjects(PyObject **mps, int n, int nadd, ...)
+static PyObject*
+multiiter_new_impl(PyObject **args)
 {
-    va_list va;
     PyArrayMultiIterObject *multi;
-    PyObject *current;
-    PyObject *arr;
-
-    int i, ntot, err=0;
+    int i;
 
-    ntot = n + nadd;
-    if (ntot < 1 || ntot > NPY_MAXARGS) {
-        PyErr_Format(PyExc_ValueError,
-                     "Need at least 1 and at most %d "
-                     "array objects.", NPY_MAXARGS);
-        return NULL;
-    }
     multi = PyArray_malloc(sizeof(PyArrayMultiIterObject));
     if (multi == NULL) {
         return PyErr_NoMemory();
     }
     PyObject_Init((PyObject *)multi, &PyArrayMultiIter_Type);
+    multi->numiter = 0;
 
-    for (i = 0; i < ntot; i++) {
-        multi->iters[i] = NULL;
-    }
-    multi->numiter = ntot;
-    multi->index = 0;
+    for (i = 0; i < NPY_MAXARGS; ++i) {
+        PyObject *obj = args[i];
+        PyObject *arr;
+        PyArrayIterObject *it;
 
-    va_start(va, nadd);
-    for (i = 0; i < ntot; i++) {
-        if (i < n) {
-            current = mps[i];
-        }
-        else {
-            current = va_arg(va, PyObject *);
-        }
-        arr = PyArray_FROM_O(current);
-        if (arr == NULL) {
-            err = 1;
+        if (obj == NULL) {
             break;
         }
-        else {
-            multi->iters[i] = (PyArrayIterObject *)PyArray_IterNew(arr);
-            if (multi->iters[i] == NULL) {
-                err = 1;
-                break;
+        if (PyObject_IsInstance(obj, (PyObject *)&PyArrayMultiIter_Type)) {
+            PyArrayMultiIterObject *mit = (PyArrayMultiIterObject *)obj;
+            int j;
+
+            if (multi->numiter + mit->numiter > NPY_MAXARGS) {
+                multiiter_wrong_number_of_args();
+                goto fail;
+            }
+            for (j = 0; j < mit->numiter; ++j) {
+                arr = (PyObject *)mit->iters[j]->ao;
+                it = (PyArrayIterObject *)PyArray_IterNew(arr);
+                if(it == NULL) {
+                    goto fail;
+                }
+                multi->iters[multi->numiter++] = it;
+            }
+        }
+        else if (multi->numiter < NPY_MAXARGS) {
+            arr = PyArray_FromAny(obj, NULL, 0, 0, 0, NULL);
+            if (arr == NULL) {
+                goto fail;
             }
+            it = (PyArrayIterObject *)PyArray_IterNew(arr);
```

Comment:
```suggestion
            it = (PyArrayIterObject *)PyArray_IterNew(arr);
            Py_DECREF(arr);
```
I think you can move the DECREF up, sorry I guess those suggestions will mess things up again... (I hope they enable suggestion to batch when in review mode...)